### DISCUSSION TITLE
Possibility to document DAG with a separated markdown file

### DISCUSSION THREAD

**Discussion by francescomucio** [apache/airflow#25396] [General]
### Description

Have the possibility to use `doc_md` also to indicate a  markdown ile containing the documentation for a DAG

### Use case/motivation

Currently [it is possible to document a DAG or task](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html#dag-task-documentation) using `doc_md`.

Sometimes these description can become quite long and pollute the DAG code, moreover it is not possible to preview the markdown description.

A possible solution would be to have the possibility to use the `doc_md` also to indicate a .md file containing the documentation for the DAG.

Bonus point if a dag (file name `my_dag.py`) is able to pick-up automatically its documentation in case a markdown file with the same name exists (e.g. `my_dag.md`). While this would be quite cool, it can appear as a sort of implicit magic, so I would like to have a few input also from the Airflow community.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
What stops you from doing:

```
doc_md = (Path(__file__).parent  /  (__file__.name + '.md')).read_text() 
```

In your DAG? This is Python. you can do that easily.

**Reply by potiuk** [reply]
And this is far more flexible than having a separate parameter.

**Reply by potiuk** [reply]
You can even write your own utility function for that : `get_dag_documentation(__file__)` or even `get_dag_documentation(__file__, dag_id)` to share it in your organization if you want to make a common convention.

**Reply by uranusjr** [reply]
`Path(__file__).with_suffix(".md").read_text()`

**Reply by potiuk** [reply]
Even simpler :)

And I even got a bit of epiphany -  I actually I think maybe simply adding the doc_md to template_fields and adding ".md" as template_ext would **just work** without adding any extra features?

**Reply by francescomucio** [reply]
yes, that was my idea, like the `sql` param for SQL operators :)

I am a simple guy, I just stitch together little things that work

**Reply by potiuk** [reply]
Yeah. So likely adding automatically "doc_md" to "templated_fields" and "template_ext" with ',md'  (if doc_md is set) should do the trick (providing that UI will read it from the rendered_templates). 

This also has nice benefit that the documentation could have jinja placeholders and the documentation COULD differe between diferent task executions - for example with different URLS etc.

But now when I am thinking - not sure if the doc_md is also read outside of the "task instance" context in the UI ? i.e. is it displayed also in  "Task View" wnen task has not been executed? Because then rendered_fields are not present.

**Reply by potiuk** [reply]
(and it might be actually more complex because doc_md is "static field" ? ) I guess?

**Reply by francescomucio** [reply]
I was thinking about doc_md at DAG level (and maybe extend to task one)

**Reply by potiuk** [reply]
Yeah. If this is not in the task, then I think the  `Path(__file__).with_suffix(".md").read_text()` seems better

**Reply by francescomucio** [reply]
Would it make sense to modify this [line](https://github.com/apache/airflow/blob/main/airflow/models/dag.py#L525) from:

```self.doc_md = doc_md```

to this:

```self.doc_md = Path(doc_md).read_text() if Path(doc_md).is_file() else doc_md ```
?

It is not templated, because it is not possible at DAG level, but it could be at task level in the way you suggested

**Reply by potiuk** [reply]
yeah. I think if that is that simple, it can be done this way.

**Reply by francescomucio** [reply]
I will give it a try

**Reply by francescomucio** [reply]
[PR](https://github.com/apache/airflow/pull/25509/files) pushed, happy to receive your feedback :)

**Comment by cvargas-xbrein** [comment]
I apologize for my English I translate it on google.
With the aim of giving my opinion and contributing, I am not an active developer, just an airflow user, I could say that in the same way as the imports are done for the .sql templates, I would like to feel the same for the documentation, and that is to say that I would like to dedicate a folder for documentation, this would imply having a .md file for dag.py but with the freedom to leave it where you want.

**Reply by potiuk** [reply]
@cvargas-geo . I think what we REALLY want to do is to make some "automated" behaviour that is simple to configure. And if you want freedom - you alreay have it.

There is nothing preventing you (even today) from doing:


```
dag.doc_md = get_dag_documentation(dag.id)
```

and your get_dag_documentation() method could look like 

```
def get_dag_documentation(id):
   return (my_local_folder / id).with_suffix(".md").read_text()
```

Easy, straightforward, super flexible, you can even enforce it via code reviews or automated CI checks.
You can apply this way any "convention" that your organisation wants to have. 
DAG is a python code. It's Imperative, i.e. you can write  whatever logic you want there. People tend to forget about it. 

Adding anything else than "automated" behavior that "just works" without defining an extra declarative way is effectively limitinbg what you can do rather than making it more flexible. You can do your own convention now, in whatever way you want. There are no limits. 

What we really came up with is that if we can do something super-simple that will not require the user to do anything but using a path rather than "content". And this is enough IMHO

**Reply by francescomucio** [reply]
@cvargas-geo this feature is merged, now you can put you documentation in a separated file, like you would do for .sql files :)

I didn't test it with a different folder, but it should work, like for the .sql files, I left the loading to jinja2 :)
